### ISSUE TITLE
Memory traffic in Formatter

### ISSUE THREAD

**Issue by StasPerekrestov** [manigandham/serilog-sinks-googlecloudlogging#30]
Hello,
I've been profiling an application I'm working on and found out that
LogFormatter causes some significant memory pressure.

![image](https://user-images.githubusercontent.com/838093/73644509-33f8fa00-467e-11ea-9036-604021add67f.png)

![image](https://user-images.githubusercontent.com/838093/73644417-07dd7900-467e-11ea-9e34-2fff8e7caf68.png)

If you don't mind I'm going to create a PR with some ideas for your review.

**Comment by manigandham** [comment]
Hi @StasPerekrestov 

Thanks for the contribution. I've updated the code further to reuse the `StringWriter` throughout the entire batch down to the log template rendering. This should eliminate many of the string allocations and improve performance. Can you test out the latest commit to see if it works for you? I'll release a new version afterwards.

**Comment by StasPerekrestov** [comment]
Thank you @manigandham 
Going to test your changes the next morning. :)

**Comment by StasPerekrestov** [comment]
Ok, some results. I haven't created an appropriate benchmark yet, but what I can see so far:
If we take MessageTemplateParser.ParseTextToken as our allocation base, then before changes we had
R = 9816700/10536766 ~= 0.93

After your recent changes this rate has become:
R = ~0.95
 which confirms that there are some improvements.

There is another source of excessive allocations and it resides in Serilog core project.
![image](https://user-images.githubusercontent.com/838093/73739875-bf8e8b80-474f-11ea-84ca-930e1b5b0c81.png)

Moreover, this 
`Serilog.Parsing.MessageTemplateParser.ParseTextToken(Int32, String, Int32)` method also performs 
lot's of Substrings.
![image](https://user-images.githubusercontent.com/838093/73739981-f49ade00-474f-11ea-9c3c-6232f499cbdc.png)

I've played a little bit with MessageTemplateParser to make it use ReadOnlyMemory<char> instead of plain String, but it may require some significant refactoring.

To sum up, thank you for your changes :)

**Comment by manigandham** [comment]
Sounds good, I'll release an update with these changes then.

**Comment by manigandham** [comment]
The changes have been released as version 2.4.0, closing this issue now.

https://www.nuget.org/packages/Serilog.Sinks.GoogleCloudLogging/2.4.0
